core/templates/dev/head/expressions/expressionSyntaxTree.js

```diff
@@ -0,0 +1,427 @@
+      };
+
+      var _verifyArgsHaveSameType = function(arg1, arg2) {
+        if (arg1 != arg2) {
+          throw new ExprWrongArgTypeError(arg1, arg1, arg2);
+        }
+        return true;
+      };
+
+      var evaluateExpression = function(expression, envs, evaluate) {
+        return evaluateParseTree(expressionParserService.parse(expression),
+          envs, evaluate);
+      };
+
+      var evaluateParseTree = function(parsed, envs, evaluate) {
+        return evaluate(parsed, envs.concat(system));
+      };
+
+      /**
+       * Looks up a variable of the given name in the env. Here the variable
+       * can be system or user defined functions and parameters, as well as
+       * system operators.
+       * @param {string} name The name to look up.
+       * @param {!Array.<!Object>} envs Represents a nested name space
+       *     environment to look up the name in. The first element is looked up
+       *     first (i.e. has higher precedence).
+       * @throws {ExprUndefinedVarError} The named variable was not found in
+       *     the given environment.
+       */
+      var lookupEnvs = function(name, envs) {
+        // Parameter value look up.
+        var value;
+        if (envs.some(function(env) {
+          if (env.hasOwnProperty(name)) {
+            value = env[name];
+            return true;
+          }
+          return false;
+        })) {
+          return value;
+        }
+
+        throw new ExprUndefinedVarError(name, envs);
+      };
+
+      // Coerces the argument to a Number, and throws an error if the result
+      // is NaN.
+      var _coerceToNumber = function(originalValue) {
+        var coercedValue = (+originalValue);
+        if (!isNaN(coercedValue)) {
+          return coercedValue;
+        }
+        throw new ExprWrongArgTypeError(
+          originalValue, typeof originalValue, 'Number');
+      };
+
+      // Coerces all values in the given argument array to Number, and throws
+      // an error if the result is NaN.
+      var _coerceAllArgsToNumber = function(args) {
+        for (var i = 0; i < args.length; i++) {
+          args[i] = _coerceToNumber(args[i]);
+        }
+        return args;
+      };
+
+      // NOTE TO DEVELOPERS: When adding a new reserved word to this object,
+      //   please first ensure that existing explorations do not use this
+      //   parameter name. Also, to prevent future explorations using it,
+      //   modify feconf.INVALID_PARAMETER_NAMES accordingly.
+      // TODO(kashida): Document all operators input and output contracts.
+
+      var system = {
+        '#': {
+          type: function(args, envs) {
+            return lookupEnvs(args + '', envs);
+          },
+          eval: function(args, envs) {
+            return lookupEnvs(args[0] + '', envs);
+          }
+        },
+        '+': {
+          type: function(args, envs) {
+            verifyNumArgs(args, 1, 2);
+            _verifyArgsHaveType(args, 'Real');
```

Comment:
Yes it does mean that. Since right now there is no real typing system for variables, we can just move everything over that needs to be a number, but it does assume that nobody has an exploration where a switches back an forth between from a being the string '3' to being a string that has characters in where we can't make a clear determination between the two types.